GH-85724: Ignore the MAC from the Touch Bar interface on macOS in uuid

ronaldoussoren opened the pull request:
All Intel MacBook Pro devices with a Touch Bar have a network interface connected to a controller for that device that has the same MAC address for all MacBooks.

Ignore these MAC adresses in the UUID library.

<!-- gh-issue-number: gh-85724 -->
* Issue: gh-85724
<!-- /gh-issue-number -->

ronaldoussoren committed:
GH-85724: Ignore the MAC from the Touch Bar interface on macOS in uuid

All Intel MacBook Pro devices with a Touch Bar have a network
interface connected to a controller for that device that has
the same MAC address for all MacBooks.

Ignore these MAC adresses in the UUID library.

vstinner commented on Lib/uuid.py:
```diff
@@ -403,9 +403,16 @@ def _get_command_stdout(command, *args):
+# MAC adresses that are blacklisted for finding a node id
+_MAC_BLACKLIST={
+    # (GH-85724) Hardcoded MAC adres for all Intel MacBook's with Touch Bar.
+    int("ac:de:48:00:11:22".replace(":", ""), 16),
```
Since there is a single MAC address, I don't think that using a set is useful here.

I suggest ``_MACBOOK_TOUCHBAR_MAC = ...``.

If you prefer a set, please rename it to "BLOCKLIST", rather than "BLACKLIST" ;-)

ronaldoussoren replied:
Naming it blacklist was lazy naming, old habits are hard to change... And as you using a set here is not useful.  Changing it to a named variable for the MAC address cleaned up other code as well.

vstinner commented on Lib/uuid.py:
```diff
@@ -403,9 +403,16 @@ def _get_command_stdout(command, *args):
 
+# MAC adresses that are blacklisted for finding a node id
+_MAC_BLACKLIST={
+    # (GH-85724) Hardcoded MAC adres for all Intel MacBook's with Touch Bar.
```
```suggestion
    # (GH-85724) Hardcoded MAC address for all Intel MacBook's with Touch Bar.
```

ronaldoussoren replied:
Sigh... If I got a euro for every time I mistype "address" I'd be rich. Guess how one spells "address" in Dutch :-)

vstinner commented on Lib/test/test_uuid.py:
```diff
@@ -830,6 +830,7 @@ def check_node(self, node, requires=None):
             print(hex, end=' ')
         self.assertTrue(0 < node < (1 << 48),
                         "%s is not an RFC 4122 node ID" % hex)
+        self.assertNotEqual(node, int("ac:de:48:00:11:22".replace(":", ""), 16))
```
Please add a comment to explain what is this MAC address.

ronaldoussoren committed:
Various cleanups

1. Remove the "BLACKLIST" variable as it is both lazy naming
   not necessary

2. Be more explicit in tests

3. Add test that checks that the mac touchbar mac is
   actually ignored.

ronaldoussoren committed:
Fix typo

ronaldoussoren commented on Lib/uuid.py:
```diff
@@ -581,6 +589,18 @@ def _netbios_getnode():
     _UuidCreate = getattr(_uuid, "UuidCreate", None)
     _has_uuid_generate_time_safe = _uuid.has_uuid_generate_time_safe
+
+    if sys.platform == "darwin" and _generate_time_safe is not None:
```
I'm not 100% sure about this test, and in particular not about the test for "darwin".  

There are people running Linux on Apple hardware and I don't know if those could be affected by this issue, in particular if there is a driver that exposes a ethernet-like interface to the Touch Bar hardware.

vstinner commented on Lib/uuid.py:
```diff
@@ -581,6 +589,18 @@ def _netbios_getnode():
+        _x = _generate_time_safe()[0]
+        if _x.endswith(bytes.fromhex(_MACOS_TOUCHBAR_MAC.replace(":", ""))):
+            _has_uuid_generate_time_safe = False
+            _generate_time_safe = None
```
Would it make sense/be possible to implement this test directly in the _uuid module: remove the function if it returns this address?

ronaldoussoren replied:
Hmmm... Possible yes, sensible maybe not.

The implementation of ``_uuid.generate_time_safe`` has 3 or 4 variants (depending on how you count) and because of that an implementation in C mirrors the Python code above but is more verbose and likely not faster than the Python implementation. 

I have a patch, but haven't tested it on a device with a Touch Bar yet. That patch adds 49 lines of code to _uuidmodule.c and removes the 4 lines of python code above.
